pkg/networkserver/mac/device_time.go

```diff
@@ -33,11 +35,27 @@ var (
 
 func HandleDeviceTimeReq(ctx context.Context, dev *ttnpb.EndDevice, msg *ttnpb.UplinkMessage) (events.Builders, error) {
 	ans := &ttnpb.MACCommand_DeviceTimeAns{
-		Time: msg.ReceivedAt,
+		Time: messageTimestamp(msg),
 	}
 	dev.MacState.QueuedResponses = append(dev.MacState.QueuedResponses, ans.MACCommand())
 	return events.Builders{
 		EvtReceiveDeviceTimeRequest,
 		EvtEnqueueDeviceTimeAnswer.With(events.WithData(ans)),
 	}, nil
 }
+
+func messageTimestamp(msg *ttnpb.UplinkMessage) *pbtypes.Timestamp {
+	var ts *pbtypes.Timestamp
+	for _, md := range msg.RxMetadata {
+		if t := md.GpsTime; t != nil {
+			return t
+		}
+		if ts == nil && md.Time != nil {
```

Comment:
This should be `md.ReceivedAt`, and the tests should be updated.

`md.Time` is the timestamp from the gateway with undefined accuracy, so using it is not ideal.